File: libs/reos_utils/lib/formatters.dart
/// Shared formatters — dates, currency, energy units.
///
/// Authority: WP-002-08 | UI/UX Design Spec v1.0 (unit/format conventions:
/// kWp, kWh). Every screen must display these value types identically.
library formatters;

import 'package:intl/intl.dart';

const String _defaultLocale = 'en_GB';

/// Format a date for display, e.g. "2 Jul 2026".
String formatDate(DateTime date, {String locale = _defaultLocale}) =>
    DateFormat('d MMM yyyy', locale).format(date);

/// Format a date with time, e.g. "2 Jul 2026, 14:30".
String formatDateTime(DateTime date, {String locale = _defaultLocale}) =>
    DateFormat('d MMM yyyy, HH:mm', locale).format(date);

/// Format a monetary amount, e.g. formatCurrency(1234.5, 'EUR') → "€1,234.50".
String formatCurrency(
  num amount,
  String currency, {
  String locale = _defaultLocale,
}) =>
    NumberFormat.currency(locale: locale, name: currency).format(amount);

/// Format installed PV capacity, e.g. formatKwp(9.87) → "9.87 kWp".
String formatKwp(num kwp, {int fractionDigits = 2}) =>
    '${kwp.toStringAsFixed(fractionDigits)} kWp';

/// Format energy, e.g. formatKwh(1234.5) → "1,234.5 kWh".
String formatKwh(
  num kwh, {
  String locale = _defaultLocale,
  int fractionDigits = 1,
}) {
  final pattern = NumberFormat.decimalPatternDigits(
    locale: locale,
    decimalDigits: fractionDigits,
  );
  return '${pattern.format(kwh)} kWh';
}

File: libs/reos-utils-ts/src/formatters.ts
/**
 * Shared formatters — dates, currency, energy units.
 *
 * Authority: WP-002-08 | UI/UX Design Spec v1.0 (unit/format conventions,
 * engineering-platform terminology: kWp, kWh).
 *
 * Every screen in the web portal and the three Flutter apps must display
 * these value types identically — do not hand-roll local formatting.
 */

const DEFAULT_LOCALE = "en-GB";

/** Format a date for display, e.g. "2 Jul 2026". */
export function formatDate(
  date: Date,
  locale: string = DEFAULT_LOCALE,
): string {
  return new Intl.DateTimeFormat(locale, {
    day: "numeric",
    month: "short",
    year: "numeric",
  }).format(date);
}

/** Format a date with time, e.g. "2 Jul 2026, 14:30". */
export function formatDateTime(
  date: Date,
  locale: string = DEFAULT_LOCALE,
): string {
  return new Intl.DateTimeFormat(locale, {
    day: "numeric",
    month: "short",
    year: "numeric",
    hour: "2-digit",
    minute: "2-digit",
  }).format(date);
}

/** Format a monetary amount, e.g. formatCurrency(1234.5, "EUR") → "€1,234.50". */
export function formatCurrency(
  amount: number,
  currency: string,
  locale: string = DEFAULT_LOCALE,
): string {
  return new Intl.NumberFormat(locale, {
    style: "currency",
    currency,
  }).format(amount);
}

/** Format installed PV capacity, e.g. formatKwp(9.87) → "9.87 kWp". */
export function formatKwp(kwp: number, fractionDigits = 2): string {
  return `${kwp.toFixed(fractionDigits)} kWp`;
}

/** Format energy, e.g. formatKwh(1234.5) → "1,234.5 kWh". */
export function formatKwh(
  kwh: number,
  locale: string = DEFAULT_LOCALE,
  fractionDigits = 1,
): string {
  const value = new Intl.NumberFormat(locale, {
    minimumFractionDigits: fractionDigits,
    maximumFractionDigits: fractionDigits,
  }).format(kwh);
  return `${value} kWh`;
}
